### ISSUE TITLE
Iterator invalidation not caught by borrow-checker

### ISSUE THREAD

**Issue by Boddlnagg** [rust-lang/rust#20232]
In the following reduced code the call to `do_bad_thing` invalidates the iterator in `main`.

```rust
fn do_bad_thing(grid: &mut Vec<Vec<u8>>) {
    grid[0] = vec![];
}

fn main() {
    let mut v: Vec<Vec<u8>> = vec![vec![1,2,3,4,5]];
    for i in v[0].iter() {
        println!("i = {}", i);
        do_bad_thing(&mut v);
    }
}
```
Running this locally produces (without optimizations):
```
i = 1
i = 2
i = 32
i = 0
i = 61
```
On playpen this does not give wrong results, but it does compile (which is the actual problem).

I'm running rustc on Windows:
```
rustc --version
rustc 0.13.0-nightly (7e11b2271 2014-12-24 20:47:12 +0000)
```

**Comment by nikomatsakis** [comment]
This may be a bug in the for loop handling code. When I do various transforms to the code, such as storing the iterator in a local, I get the expected error messages.

**Comment by eddyb** [comment]
@nikomatsakis I found some of those a long while ago (#17068), guess there's more?
Wait, no the test case above has no `for` loops at all!

**Comment by pythonesque** [comment]
@nikomatsakis It's not related to `for`.

**Comment by nikomatsakis** [comment]
Fix pending.

**Comment by nikomatsakis** [comment]
See PR #20751

**Comment by alexcrichton** [comment]
Fixed in https://github.com/rust-lang/rust/pull/20751
